integrations/jta/jdbc/src/main/java/io/helidon/integrations/jta/jdbc/LocalXAResource.java

```diff
@@ -0,0 +1,1031 @@
+        //
+        // Finally, if an error happens during commit, we issue a SQL/JDBC/local transaction rollback() command.  If
+        // that works, then we return XA_RB*. If that fails, chances are we return XAER_RMERR, but
+        this.computeAssociation(XARoutine.COMMIT,
+                                xid,
+                                EnumSet.of(Association.BranchState.IDLE,
+                                           Association.BranchState.PREPARED),
+                                a -> commitAndReset(a, onePhase),
+                                true); // do remove association on error
+
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.exiting(this.getClass().getName(), "commit");
+        }
+    }
+
+    @Override // XAResource
+    public void rollback(Xid xid) throws XAException {
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.entering(this.getClass().getName(), "rollback", xid);
+        }
+        requireNonNullXid(xid);
+
+        // An error during rollback is bad; in a two-phase situation where we've already prepared, then returning
+        // XAER_RMERR will put us in HEURISTIC_HAZARD
+        // (https://github.com/jbosstm/narayana/blob/c5f02d07edb34964b64341974ab689ea44536603/ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/resources/arjunacore/XAResourceRecord.java#L379-L420).
+        // Doing XAER_RMFAIL will put us in FINISH_ERROR.
+        this.computeAssociation(XARoutine.ROLLBACK,
+                                xid,
+                                EnumSet.of(Association.BranchState.IDLE,
+                                           Association.BranchState.PREPARED,
+                                           Association.BranchState.ROLLBACK_ONLY),
+                                LocalXAResource::rollbackAndReset,
+                                true); // do remove association on error
+
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.exiting(this.getClass().getName(), "rollback");
+        }
+    }
+
+    @Override // XAResource
+    public void forget(Xid xid) throws XAException {
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.entering(this.getClass().getName(), "forget", xid);
+        }
+        requireNonNullXid(xid);
+
+        this.computeAssociation(XARoutine.FORGET,
+                                xid,
+                                EnumSet.of(Association.BranchState.HEURISTICALLY_COMPLETED),
+                                LocalXAResource::forgetAndReset,
+                                false); // don't remove association on error
+
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.exiting(this.getClass().getName(), "forget");
+        }
+    }
+
+    @Override // XAResource
+    public Xid[] recover(int flags) throws XAException {
```

Comment:
I believe this method may be invoked by any thread at any time at the discretion of the transaction manager to recover from a failure with the two-phase completion process (so something going wrong after the `start`/`end` cycle).  It is sort of irrelevant as this "local" `XAResource` implementation can't recover things anyway.